### PR TITLE
PLAN-42-avatar-upload

### PR DESCRIPTION
**Pull request by AndzejIlj**

### CODE REVIEW

File: services/api/src/domain/player.ts [hunk 1]
```diff
@@ -13,6 +13,7 @@ export interface Player {
   id: string;
   name: string;
   color: PlayerAvatarColor;
+  avatar: string;
```
Comment (jvalatkevicius) [inline services/api/src/domain/player.ts:16]: is this mandatory or optional?

File: services/api/src/messaging/players.service.ts [hunk 2]
```diff
@@ -157,9 +160,10 @@ export const successfullyJoined = (
 
 export const playerConnected: Handler = (
   ws,
-  { playerName, id, role }: PlayerConnectedMessage,
+  { playerName, id, role, avatar }: PlayerConnectedMessage,
 ): void => {
   const newPlayerId = id ? id : uuidv4();
+  const avatarUrl = avatar ? avatar : '';
```
Comment (jvalatkevicius) [inline services/api/src/messaging/players.service.ts:166]: Why not use `undefined` here?

File: services/app/src/api/playerApi.js [hunk 3]
```diff
@@ -5,12 +5,17 @@ export const buildMoveSkippedEventMessage = (playerId) => ({
...
 });
 
 export const buildPlayerFigureMovedMessage = (turn) => ({
   type: WS_EVENT_TYPES.FIGURE_MOVED,
   payload: turn,
-});
\ No newline at end of file
+});
+
+export const buildPlayerAvatarMessage = (url) => ({
```
Comment (jvalatkevicius) [inline services/app/src/api/playerApi.js:18]: `buildPlayerAvatarUpdateMessage`

File: services/app/src/components/avatarPicture/CustomAvatar.jsx [hunk 4]
```diff
@@ -0,0 +1,51 @@
+import React, { useState } from 'react';
```
Comment (jvalatkevicius) [inline services/app/src/components/avatarPicture/CustomAvatar.jsx:1]: `avatarPicture` is excessive. `avatar` is enough

File: services/app/src/components/avatarPicture/CustomAvatar.jsx [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+        size={size}
+        id={id}
+        isBorderNeeded={isBorderNeeded}
+      />  
+    )
+  }
+
+  return (
+    <Avatar 
+      size={getAvatarPxSize(size)}
```
Comment (jvalatkevicius) [inline services/app/src/components/avatarPicture/CustomAvatar.jsx:22]: why this `size` property is wrapped in `getAvatarPxSize` while the `DefaultAvatar` `size` property is not?

File: services/app/src/components/avatarPicture/CustomAvatar.jsx [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+      style={{ backgroundColor: '#F3F6FA', border: '1px solid var(--primary)' }}
```
Comment (jvalatkevicius) [inline services/app/src/components/avatarPicture/CustomAvatar.jsx:23]: why inline styles?

File: services/app/src/components/avatarPicture/GhostAvatar.jsx [hunk 7]
```diff
@@ -0,0 +1,34 @@
+import React from 'react'
+import PropTypes from 'prop-types';
+import { Avatar } from 'antd';
+import { getAvatarPxSize, getAvatarFontSize } from '../../helpers/getAvatarSizes';
+
+const GhostAvatar = ({ avatarText, size }) => (
+  <Avatar 
+    style={{
```
Comment (jvalatkevicius) [inline services/app/src/components/avatarPicture/GhostAvatar.jsx:8]: why inline styles?

File: services/app/src/components/avatarPicture/DefaultAvatar.jsx [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+  const border = isBorderNeeded 
+    ? `1px solid ${rgbToColor(user.color.text)}` 
+    : `1px solid ${rgbToColor(user.color.background)}`;
+
+  if (user) {
+    return (
+      <Avatar
+        size={getAvatarPxSize(size)} 
+        style={{
```
Comment (jvalatkevicius) [inline services/app/src/components/avatarPicture/DefaultAvatar.jsx:20]: why inline styles?

File: services/app/src/components/avatarPicture/CustomAvatar.jsx [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+  )
+}
+
+CustomAvatar.defaultProps = {
+  size: 'medium',
+  isBorderNeeded: false,
+}
+
+CustomAvatar.propTypes = {
+  size: PropTypes.oneOf(['x-small', 'small', 'medium', 'large']),
```
Comment (jvalatkevicius) [inline services/app/src/components/avatarPicture/CustomAvatar.jsx:45]: move prop type into file where it could be reused.
size values should match the naming convetion we use for spacing (xs, s, m, ...)

File: services/app/src/helpers/getAvatarSizes.js [hunk 10]
```diff
@@ -0,0 +1,27 @@
+export const getAvatarPxSize = (size) => {
```
Comment (jvalatkevicius) [inline services/app/src/helpers/getAvatarSizes.js:1]:
```
const sizesMap = {
  'xs': 26,
  's': 36,
  ...
}

const fontSizesMap = {
  'xs': 14,
  's': 18,
  ...
}

```

File: services/app/src/helpers/getAvatarSizes.js [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+    return 36;
+  }
+  if (size === 'large') {
+    return 96;
+  }
+
+  return 40;
+}
+
+export const getAvatarFontSize = (size) => {
```
Comment (jvalatkevicius) [inline services/app/src/helpers/getAvatarSizes.js:15]: Why this returns string in pixels, while the one above returns number?

File: services/app/src/components/avatarUploadModal/AvatarModalContainer.jsx [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+          imageUrl={imageUrl} 
+          retryPictureUpload={retryPictureUpload} 
+          handleOkBtnPress={handleOkBtnPress} 
+        /> 
+      )}
+    </Modal>
+  )
+}
+
+AvatarUploadModal.defaultProps = {
```
Comment (jvalatkevicius) [inline services/app/src/components/avatarUploadModal/AvatarModalContainer.jsx:77]: ?

File: services/app/src/components/avatarUploadModal/AvatarModalContainer.jsx [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+import PropTypes from 'prop-types';
+import { Modal } from 'antd';
+import './avatarModal.css';
+import AvatarModalForm from './AvatarModalForm';
+import AvatarModalComplete from './AvatarModalComplete';
+import { useUserContext } from '../../contexts/UserContext';
+import { buildPlayerAvatarMessage } from '../../api/playerApi';
+import { useWsContext } from '../../contexts/ws-context';
+
+const AvatarUploadModal = ({ showAvatarModal, setShowAvatarModal }) => {
```
Comment (jvalatkevicius) [inline services/app/src/components/avatarUploadModal/AvatarModalContainer.jsx:11]: File name does not match the component name

File: services/app/src/components/avatarUploadModal/avatarModal.css [hunk 14]
```diff
@@ -0,0 +1,14 @@
+.avatar-modal {
+  max-width: 600px;
+  padding: 20px;
```
Comment (jvalatkevicius) [inline services/app/src/components/avatarUploadModal/avatarModal.css:3]: spacing.css

File: services/app/src/components/avatarUploadModal/AvatarModalContainer.jsx [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+      title="Upload profile picture"
+      open={showAvatarModal}
+      className='avatar-modal'
+      footer={null}
+      onCancel={closeAvatarModule}
+    >
+      {modalStep === 1 && (
+        <AvatarModalForm 
+          handleOkBtnPress={handleOkBtnPress} 
+          urlText={urlText}
```
Comment (jvalatkevicius) [inline services/app/src/components/avatarUploadModal/AvatarModalContainer.jsx:61]: `url` is enough

File: services/app/src/components/avatarUploadModal/AvatarModalContainer.jsx [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+      open={showAvatarModal}
+      className='avatar-modal'
+      footer={null}
+      onCancel={closeAvatarModule}
+    >
+      {modalStep === 1 && (
+        <AvatarModalForm 
+          handleOkBtnPress={handleOkBtnPress} 
+          urlText={urlText} 
+          setUrlText={setUrlText}
```
Comment (jvalatkevicius) [inline services/app/src/components/avatarUploadModal/AvatarModalContainer.jsx:62]: same as previous

File: services/app/src/components/avatarPicture/DefaultAvatar.jsx [hunk 17]
```diff
@@ -0,0 +1,50 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { Avatar } from 'antd';
+import { getAvatarPxSize, getAvatarFontSize } from '../../helpers/getAvatarSizes';
+import { useChessBoardContext } from '../../contexts/ChessBoardContext';
+import { rgbToColor } from '../../helpers/rgbToColor';
+
+const DefaultAvatar = ({ size, id, isBorderNeeded }) => {
+  const { findUserById } = useChessBoardContext();
```
Comment (jvalatkevicius) [inline services/app/src/components/avatarPicture/DefaultAvatar.jsx:9]: Check if you could use currentplayer instead of passing id

File: services/app/src/components/avatarPicture/GhostAvatar.jsx [hunk 18]
```diff
@@ -0,0 +1,34 @@
+import React from 'react'
+import PropTypes from 'prop-types';
+import { Avatar } from 'antd';
+import { getAvatarPxSize, getAvatarFontSize } from '../../helpers/getAvatarSizes';
+
+const GhostAvatar = ({ avatarText, size }) => (
+  <Avatar
```
Comment (jvalatkevicius) [inline services/app/src/components/avatarPicture/GhostAvatar.jsx:7]: Make a wrapper Avatar component, that would wrap antd's Avatar and apply the styling inside the wrapper to avoid copy-pasting the same thing.

File: services/app/src/components/chessBoard/Square.jsx [hunk 19]
```diff
@@ -13,45 +14,70 @@ const Square = ({
...
+          {avatarToRender.map((player, index) => (
+            <UserAvatar
+              isBorderNeeded
+              size='x-small' 
+              id={player.playerId} 
+              key={`bubble-${index}`} 
+              avatarText={player.player[0]}
+            />
+          ))}
+          <Avatar
```
Comment (jvalatkevicius) [inline services/app/src/components/chessBoard/Square.jsx:39]: try using antd's avatar group for this funcitonality

File: services/app/src/components/chessBoard/Square.jsx [hunk 20]
```diff
@@ -69,7 +95,7 @@ const Square = ({
     >
       {!!items.length && <span className={classNames(["number", "number-row", filled && "number-filled"])}>{board[row][0].attribute}</span>}
       <div className="bubble-container">
-        {items.map(renderBubble)}
+        {renderBubble(items)}
```
Comment (jvalatkevicius) [inline services/app/src/components/chessBoard/Square.jsx:98]: make it look like this
```
      {!!items.length && <span className={classNames(["number", "number-row", filled && "number-filled"])}>{board[row][0].attribute}</span>}
      <div className="bubble-container">
        <Avatar.Group maxCount={3}>
          {items.map((player, index) => (
            <UserAvatar 
              size='x-small' 
              id={player.playerId} 
              key={`bubble-${index}`}
              avatarText={player.player[0]}
              isBorderNeeded
            />
          ))}
        </Avatar.Group>
      </div>
```

File: services/app/src/components/avatarPicture/GhostAvatar.jsx [hunk 21]
```diff
@@ -0,0 +1,34 @@
+import React from 'react'
+import PropTypes from 'prop-types';
+import { Avatar } from 'antd';
+import { getAvatarPxSize, getAvatarFontSize } from '../../helpers/getAvatarSizes';
+
+const GhostAvatar = ({ avatarText, size }) => (
```
Comment (jvalatkevicius) [inline services/app/src/components/avatarPicture/GhostAvatar.jsx:6]: Ghost naming is used to describe button styles in antd, please think of another name for this kind of avatar

File: services/app/src/components/avatarPicture/DefaultAvatar.jsx [hunk 22]
```diff
@@ -0,0 +1,50 @@
...
+  const user = findUserById(id);
+
+  const border = isBorderNeeded 
+    ? `1px solid ${rgbToColor(user.color.text)}` 
+    : `1px solid ${rgbToColor(user.color.background)}`;
+
+  if (user) {
```
Comment (jvalatkevicius) [inline services/app/src/components/avatarPicture/DefaultAvatar.jsx:16]: why is this check needed?

File: services/app/src/static/style/antd-override.css [hunk 23]
```diff
@@ -1,7 +1,55 @@
+/*Buttons*/
+.ant-btn-primary:disabled:hover {
+  background-color: rgba(0, 0, 0, 0.04) !important;
```
Comment (jvalatkevicius) [inline services/app/src/static/style/antd-override.css:3]: Please check antd `ConfigProvider` and make antd use our primary and secondary colors

**Review by jvalatkevicius — CHANGES_REQUESTED** [review]

File: services/api/src/domain/player.ts (hunk 1 above)
Comment (AndzejIlj) [reply services/api/src/domain/player.ts:16]: I would say it is mandatory. If the user does not have custom avatar, its gonna be '' (empty string)

File: services/app/src/components/avatarPicture/DefaultAvatar.jsx (hunk 17 above)
Comment (AndzejIlj) [reply services/app/src/components/avatarPicture/DefaultAvatar.jsx:9]: No, we can not, because this component also renders team's avatars, so currentPlayerId would not work in this case

File: services/app/src/helpers/getAvatarSizes.js (hunk 11 above)
Comment (AndzejIlj) [reply services/app/src/helpers/getAvatarSizes.js:15]: Because antd avatar takes size prop in integer values and font size goes to style attribute, so it needs to be a string

File: services/app/src/components/avatarPicture/CustomAvatar.jsx (hunk 5 above)
Comment (AndzejIlj) [reply services/app/src/components/avatarPicture/CustomAvatar.jsx:22]: If you are talking about this particular file, then it is not wrapped in ```getAvatarPxSize``` because ```DefaultAvatar``` takes size as a prop(small, medium, ect). And then in ```DefaultAvatar.jsx``` it takes that argument and calls it with ```getAvatarPxSize```.

File: services/app/src/components/gameHeaderDropDown/GameHeaderDropDown.jsx [hunk 24]
```diff
@@ -0,0 +1,51 @@
...
+    if (open) {
+      setImgSrc(DropdownArrowUp);
+      return;
+    }
+    setImgSrc(DropdownArrowDown);
+  }
+
+  const handleButtonClick = (callbackFn) => {
+    setImgSrc(DropdownArrowDown);
+    callbackFn();
```
Comment (jvalatkevicius) [inline services/app/src/components/gameHeaderDropDown/GameHeaderDropDown.jsx:21]: why is callback needed here?

File: services/app/src/components/gameHeaderDropDown/GameHeaderDropDown.jsx [hunk 25]
```diff
@@ -0,0 +1,51 @@
...
+
+  const rotateArrow = (open) => {
+    if (open) {
+      setImgSrc(DropdownArrowUp);
+      return;
+    }
+    setImgSrc(DropdownArrowDown);
+  }
+
+  const handleButtonClick = (callbackFn) => {
```
Comment (jvalatkevicius) [inline services/app/src/components/gameHeaderDropDown/GameHeaderDropDown.jsx:19]: think of better name

File: services/app/src/components/loginPage/LoginForm.jsx [hunk 26]
```diff
@@ -49,10 +61,27 @@ const LoginForm = () => {
         </div>
       </div>
 
+      <div className='form-input-container margin-sm'>
```
Comment (jvalatkevicius) [inline services/app/src/components/loginPage/LoginForm.jsx:64]: lets us antd's form here

File: services/app/src/helpers/getAntdTheme.js [hunk 27]
```diff
@@ -0,0 +1,5 @@
+export const getAntdTheme = () => ({
```
Comment (jvalatkevicius) [inline services/app/src/helpers/getAntdTheme.js:1]: why is this a function?

**Review by jvalatkevicius — CHANGES_REQUESTED** [review]

File: services/api/src/messaging/players.service.ts [hunk 28]
```diff
@@ -157,9 +160,10 @@ export const successfullyJoined = (
 
 export const playerConnected: Handler = (
   ws,
-  { playerName, id, role }: PlayerConnectedMessage,
+  { playerName, id, role, avatar }: PlayerConnectedMessage,
 ): void => {
   const newPlayerId = id ? id : uuidv4();
+  const avatarUrl = avatar ? avatar : undefined;
```
[the remark this answers is not in the document]
Comment (AndzejIlj) [reply services/api/src/messaging/players.service.ts:166]: Yes, it is, because without this line, the player object get an avatar property with value of null. And I think we don't want that, if the player does not have an avatar, he/she should not contain an avatar property

File: services/app/src/components/avatarUploadModal/AvatarModalForm.jsx [hunk 29]
```diff
@@ -0,0 +1,85 @@
...
+
+  useEffect(() => {
+    userContext.setAvatarError(false);
+
+    if(urlInputRef.current) {
+      urlInputRef.current.focus();
+    }
+  }, []);
+
+  useEffect(() => {
```
Comment (jvalatkevicius) [inline services/app/src/components/avatarUploadModal/AvatarModalForm.jsx:31]: `useMemo`

File: services/app/src/components/avatarUploadModal/AvatarUploadModal.jsx [hunk 30]
```diff
@@ -0,0 +1,87 @@
...
+    setImageUrl('');
+    userContext.setAvatarError(false);
+  }
+
+  const moveToFinalStep = (url) => {
+    try {
+      const urlHref = new URL(url).href;
+      setImageUrl(urlHref);
+      setModalStep(2);
+    } catch {
```
Comment (jvalatkevicius) [inline services/app/src/components/avatarUploadModal/AvatarUploadModal.jsx:30]: Why is this needed? There are not async calls, what can throw here?

File: services/app/src/components/avatarUploadModal/AvatarModalForm.jsx [hunk 31]
```diff
@@ -0,0 +1,85 @@
...
+      return;
+    }
+    setIsPrimaryBtnDisabled(true);
+  }, [urlValue]);
+  
+  
+  return (
+    <>
+      <p className='avatar-upload-description font-size-s margin-b-l'>Enter a link of the image</p>
+      <Form
```
Comment (jvalatkevicius) [inline services/app/src/components/avatarUploadModal/AvatarModalForm.jsx:43]: have a look at `Form`s docs, especially how `onSubmit` works.

File: services/app/src/components/avatarUploadModal/AvatarUploadModal.jsx [hunk 32]
```diff
@@ -0,0 +1,87 @@
...
+  const userContext = useUserContext();
+ 
+  const [modalStep, setModalStep] = useState(1);
+  const [imageUrl, setImageUrl] = useState('');
+
+  const onAvatarModalCancel = () => {
+    setModalStep(1);
+    closeAvatarModal();
+    setImageUrl('');
+    userContext.setAvatarError(false);
```
Comment (jvalatkevicius) [inline services/app/src/components/avatarUploadModal/AvatarUploadModal.jsx:22]: do we really need to move avatar error into user context? It seems like something that's localized within this component's scope and should be a state here.

File: services/app/src/pages/game/Game.jsx [hunk 33]
```diff
@@ -62,7 +66,11 @@ const Game = () => {
   return (
     <>
       {isAnotherSessionActive && <Navigate to={ROUTES.userTaken} />}
-      <GameHeader />
+      <AvatarUploadModal 
+        showAvatarModal={showAvatarModal}
+        closeAvatarModal={() => setShowAvatarModal(false)}
```
Comment (jvalatkevicius) [inline services/app/src/pages/game/Game.jsx:71]: rename prop to `onClose` you do not need to specify what modal are you closing when passing prop to `AvatarUploadModal` component

File: services/app/src/pages/game/Game.jsx [hunk 34]
```diff
@@ -62,7 +66,11 @@ const Game = () => {
   return (
     <>
       {isAnotherSessionActive && <Navigate to={ROUTES.userTaken} />}
-      <GameHeader />
+      <AvatarUploadModal 
+        showAvatarModal={showAvatarModal}
```
Comment (jvalatkevicius) [inline services/app/src/pages/game/Game.jsx:70]: rename this to `isVisible` or `isOpen`

File: services/app/src/components/avatarUploadModal/AvatarModalForm.jsx [hunk 35]
```diff
@@ -0,0 +1,85 @@
...
+    if (e.code === 'Enter') {
+      moveToFinalStep(urlValue);
+    }
+  }
+
+  useEffect(() => {
+    userContext.setAvatarError(false);
+
+    if(urlInputRef.current) {
+      urlInputRef.current.focus();
```
Comment (jvalatkevicius) [inline services/app/src/components/avatarUploadModal/AvatarModalForm.jsx:27]: have you tried `autoFocus` property on input?

**Review by jvalatkevicius — CHANGES_REQUESTED** [review]

File: services/app/src/components/avatarUploadModal/AvatarUploadModal.jsx (hunk 30 above)
Comment (AndzejIlj) [reply services/app/src/components/avatarUploadModal/AvatarUploadModal.jsx:30]: Without the catch statement, if the ```new URL``` fails to construct a valid object, the error gets printed in the `console.log` but the program does not crash. I thought it shouldn't so I wrapped it in try catch statement.
Removed it.

File: services/app/src/components/avatarUploadModal/AvatarModalForm.jsx (hunk 35 above)
Comment (AndzejIlj) [reply services/app/src/components/avatarUploadModal/AvatarModalForm.jsx:27]: Yes I have, for some reason it does not work. I think its because the input is wrapped in antD's form.

Comment (AndzejIlj) [reply services/app/src/components/avatarUploadModal/AvatarModalForm.jsx:27]: Upadte: I saw, that it works in login page, but does not work in this modal, so I did a little research and cakme to an answer that autofocus does not work inside a modal. HTML `autoFocus` means focus input when the page loads. The solution is useRef. 
Source: https://github.com/ant-design/ant-design/issues/8668

File: services/app/src/components/avatar/CustomAvatar.jsx [hunk 36]
```diff
@@ -0,0 +1,60 @@
...
+      size={avatarSizesMap[size]}
+      style={getCustomAvatarStyle()}
+      src={(
+        <img 
+          src={imageUrl}
+          alt='profile pic'
+          onError={() => {
+            setIsAvatarError(true);
+            if (playerId === currentPlayerId) {
+              onError();
```
Comment (jvalatkevicius) [inline services/app/src/components/avatar/CustomAvatar.jsx:36]: this prop should be optional and called conditionally
`onError?.()`

File: services/app/src/components/avatar/CustomAvatar.jsx [hunk 37]
```diff
@@ -0,0 +1,60 @@
...
+    <Avatar 
+      size={avatarSizesMap[size]}
+      style={getCustomAvatarStyle()}
+      src={(
+        <img 
+          src={imageUrl}
+          alt='profile pic'
+          onError={() => {
+            setIsAvatarError(true);
+            if (playerId === currentPlayerId) {
```
Comment (jvalatkevicius) [inline services/app/src/components/avatar/CustomAvatar.jsx:35]: why?

File: services/app/src/components/avatarUploadModal/AvatarUploadModal.jsx [hunk 38]
```diff
@@ -0,0 +1,86 @@
...
+  
+  const onAvatarModalCancel = () => {
+    setModalStep(1);
+    onClose();
+    setImageUrl('');
+    setIsModalAvatarError(false);
+  }
+
+  const moveToFinalStep = (url) => {
+    const urlHref = new URL(url).href;
```
Comment (jvalatkevicius) [inline services/app/src/components/avatarUploadModal/AvatarUploadModal.jsx:28]: lets remove this

**Review by jvalatkevicius — CHANGES_REQUESTED** [review]

**Review by jvalatkevicius — CHANGES_REQUESTED** [review]

File: services/app/src/components/gameHeaderDropDown/dropdown.css [hunk 39]
```diff
@@ -0,0 +1,3 @@
+.dropdown-icon:hover {
```
Comment (jvalatkevicius) [inline services/app/src/components/gameHeaderDropDown/dropdown.css:1]: is this needed?

**Review by jvalatkevicius — CHANGES_REQUESTED** [review]

**Review by jvalatkevicius — DISMISSED** [review]

**Review by jvalatkevicius — DISMISSED** [review]

**Review by jvalatkevicius — APPROVED** [review]
